File: Solutions/Python/take_gifts_from_the_richest_pile.py
"""
    [EASY]
    2558. Take Gifts From the Richest Pile

    Concepts:
    - heap/priority queue

    Stats:
        Runtime | 3 ms      [Beats 87.72%]
        Memory  | 17.39 MB  [Beats 20.60%]
"""

import heapq

class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        gifts = [-gift for gift in gifts]
        heapq.heapify(gifts)
        while k:
            heapq.heapreplace(gifts, -floor(sqrt(-gifts[0])))
            k -= 1
        
        return -sum(gifts)
            

File: Solutions/Cpp/take_gifts_from_the_richest_pile.cpp
/*
    [EASY]
    2558. Take Gifts From the Richest Pile

    Concepts:
    - heap/priority queue

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 13.60 MB  [Beats 7.74%]
*/

class Solution {
public:
    long long pickGifts(vector<int>& gifts, int k) {
        std::priority_queue<long> pq(gifts.begin(), gifts.end());
        long long sum = 0;
        int val = 0;

        while (k != 0) {
            val = pq.top();
            pq.pop();
            pq.push(std::floor(std::sqrt(val)));
            k--;
        }

        for (; !pq.empty(); pq.pop())
            sum += pq.top();
        return sum;
    }
};